### ISSUE TITLE
CorrelationContext

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#344]
I did a short test using the principal names as endpoint identity for DTLS and notifies.

This extents the exporting of the `CorrelationContext` over the scope of the implementation of `ObservationStore`. So I would like to have the usage and expectation more defined.

FMPOV a CorrelationContext should be considered as a "unmodifiable map of string to string". The current used types (e.g. `DtlsCorrelationContext`) should not be mandatory, I would even like to change them to Builders. That would remove some comfort methods (e.g.  `DtlsCorrelationContext.getEpoch()` ) but on the other side it would simplify to provide a CorrelationContext as input for requests and the save/restore mechanism in the `ObservationStore`.

So any thoughts about that?

**Comment by sbernard31** [comment]
> I did a short test using the principal names as endpoint identity for DTLS and notifies.

What do you mean by _for DTLS_ ?

**Comment by boaks** [comment]
"principal names" could only identify a endpoint using DTLS, as far as I know.
And for the normal request/response mapping, in my opinion the currently used address:port mapping should do it. So the "main pain" with the ip-address changes right now seems to be the notifies. Therefore I wrote "endpoint identity for DTLS and notifies".

**Comment by sbernard31** [comment]
About request/response mapping, I never experiment it but I suspect that with NAT timeout/retransmission/deduplication you could sometime face some unexpected behavior.

But I agree the priority is observation, so let's focus on it.

It seems that your proposition target the same issue than #322.
We start a discussion about that there. I just missed the last response (sry about that). Should we continue the discussion to find the best solution with what we have in mind  here or at #322 ?

**Comment by boaks** [comment]
> Should we continue the discussion to find the best solution with what we have in mind here or at #322 ?

Where ever you prefer.

But, let me just say, I think, that independent from the usage in #322, I think, if we use the correlation context  "external", as a external `ObservationStore` implementation would also imply, we should define some more about the rules to use it.

**Comment by sbernard31** [comment]
> if we use the correlation context "external", as a external ObservationStore implementation would also imply, we should define some more about the rules to use it.

you mean for example limit it to an unmodifiable map of string to string?
For the unmodifiable map, I agree. 
For the string to string, I could agree. Data in correlation constraint will probably be serialized/deserialized at some point or another, So I ask myself if we should do that when we create the context (so using string=>string) or when we really need to serialize/deserialize it (and using string => object)?

---------
I have concerns about using correlation context as identifier instead of endpoint identifier for use cases where peer don't use authentication, or where the identifier chosen could not be know by the user before. (I mean cannot be attached to the request by the user)
I know you think this is a `NONE` use case, but in this case the full correlation context can not be used as identifier so you need a way to sort which part of the context you keep as identifier.

The other concern about using correlation as identifier instead of endpoint identifier, is for the use case below:

"_When I receive a request from an endpoint,
 I want to send an observe request to this endpoint._"

If we use endpoint identifier, it's easy : 
I receive a request from an endpoint with a given identifier.
I send a request to this endpoint by setting this given identifier.

With correlation context, not so easy : 
I receive a request with the full correlation context (epoch cipher suite ..)
Then when I want to send the request back, I need to create a new context and only set what I need.

**Comment by boaks** [comment]
> With correlation context, not so easy :
> I receive a request with the full correlation context (epoch cipher suite ..)
> Then when I want to send the request back, I need to create a new context and only set what I need.

Why?

> I receive a request with the full correlation context (epoch cipher suite ..  principal name)

So it's up to my matcher, how it uses the other fields, if the "endpoint" (principal name) is provided.

**Comment by sbernard31** [comment]
because you don't want to use the full correlationcontext as observation identifier. (you probably want just the identity or just the sessionID, or ... )

**Comment by boaks** [comment]
> because you don't want to use the full correlationcontext as observation identifier. (you probably want just the identity or just the sessionID, or ... )

But,if we pass the correlation context as additional parameter in the ObservationStore, would this leave it to the ObservationStore implementation, what is chosen as additional id (additional to the token)?

If you setup californium to match related to the endpoint id, then you may also provide an ObservationStore, which use that endpoint id.

I currently try to isolate my changes for demonstrating the use of such a correlation context extension. I did this last week and found, that with the current `DtlsCorrelationContext` this would end up in a lot of changes just for adjust the test usage of that. So currently I try to prepare a demonstration with an additional constructor for a  `DtlsCorrelationContext`, which should result in a more reasonable number of changes ...

**Comment by sbernard31** [comment]
For observationStore, this can maybe make the tricks like that but this means
- there is 2 classes which sort what is the important "things" in the context, `observationStore` and `correlationMatcher`) 
- this doesn't cover the more general use case request/response matching.

**Comment by boaks** [comment]
> this doesn't cover the more general use case request/response matching.

The general use case for request/response matching should in my opinion use the ip address, because the timespan should be very short.

So only notifies for none DTLS (or none unique credentials) will then be not covered. 
But I didn't see/understand, on which protocol based artifacts a other long term identity should be based on. May be the IETF/core introduce such an artifact.

With some other current discussions:
- generate tokens generally unique
- correlation context matcher replaceable

I guess, it would be simply possible to build a "unsecure" and dangerous version, where only the token is used to match notifies. I hope, that those who does this, will document there decisions.

**Comment by boaks** [comment]
PR #350 should give a first impression, how the correlation context extension would work.

**Comment by sbernard31** [comment]
> So only notifies for none DTLS (or none unique credentials) 

We can put this use-case aside.   

>The general use case for request/response matching should in my opinion use the ip address, because the timespan should be very short.

As I said before, I have concern about NAT environment, retransmission and deduplication but nothing concrete.

Using identifier request/response by peer can help for token generation too. (see #173)

**Comment by boaks** [comment]
I assume that request/response should be within the NAT timeout. If not, I agree, it would not work.

But as long as you don't define, on which protocol artifact you will base your identifier, I can't see, that there is a solution.

**Comment by boaks** [comment]
https://github.com/eclipse/californium/issues/344#issuecomment-313108049

> you mean for example limit it to an unmodifiable map of string to string?
> For the unmodifiable map, I agree.
> For the string to string, I could agree. Data in correlation constraint will probably be serialized/deserialized at some point or another, So I ask myself if we should do that when we create the context (so using string=>string) or when we really need to serialize/deserialize it (and using string => object)?

Therefore I would change the "specification" explicitly to a unmodifiable map and then reuse them.
Using `Object` would require to have a serialization. Depending on the technology, this may not be generic, and so may cause trouble. Or what would you propose for such Objects?

**Comment by sbernard31** [comment]
About serialization nothing more than what you propose (String serialization).
I just ask myself when this serialization should be done.

**Comment by boaks** [comment]
OK. 
So one part of my idea was, if it's unmodifiable, we can keep and reuse it. 
Therefore the penalty of the "string serialization" should be limited. 
Using `String` instead of `Object `would indicate, that `toString()` is required and e.g. for `X509CertPath` it would show more explicit, that a useful `toString()` is required.

**Comment by sbernard31** [comment]
In my branch about endpoint identifier I use object to let the serialization question for later. I think we don't need to agree on that now. Moving from one to another should not be so hard. And personally I don't have strong opinion about that.

The endpoint identifier or correlationContext question seems more impacting.

**Comment by boaks** [comment]
After a closer look to your PR #322 my conclusion is:

Both approaches (PR #322 and #350) requires to add something as "endpoint identifier" to the correlation context.
The one approach takes Object and a conversion-factory (which implementation must match to the used Object type in the correlation context).
The other defines, that the "endpoint identifier" is a "String" and force the connector to provided that.

But I don't see, that the conversion-factory generally offers more functionality to handle endpoint identities, it rely on information added to the correlation context.

**Comment by sbernard31** [comment]
I agree that both approaches do approximately the same thing this is my point in my first comment.

**I think the needs is :**
1) make visible an endpoint identifier concept at CoAP level. I mean we talk about that regularly in our discussions for different purpose/feature.
2) what is used as identifier should be configurable/customizable. (because of incertitude about the spec and also maybe to be more flexible)

**Why I prefer the Identifier/Factory approach ?**
   a) From my point of view if we have an identifier, CorrelationContext is no more needed at CoAP level. The identifier is the only thing we care. Keeping CorrelationContext at connector level can make sense. I would even call this ConnectorContext.
   b) The way we translate identifier to context and vice/versa, is clearly defined and centralized in one class.
   c) It's easy to change the way we identify an endpoint, just change this factory.

**What could be improved in the Identifier/Factory approach ?** Serialization/Deserialization we can imagine to add this kind of method on EndpointIdentifier interface. 

I will do some comments about #350. To complete this post.

**Comment by boaks** [comment]
So, what's about:
- move `principal `from `Request `to `Message`, so it could also passed in.
- use then the `Message `as "identifier" for get/remove (and implicit for add) in the `ObservationStore`.

This will only slightly change the existing code and offers the implementation flexibility for observe/notify by implementing the "endpoint-id/factory" there or just use "getName()".
For that I would go for extend the "Matcher" for explicitly matching a notify.

**Comment by sbernard31** [comment]
> "use then the Message as "identifier" for get/remove (and implicit for add) in the ObservationStore"_

This seems strange to me, It make the identifier endpoint even less visible than when we use "CorrelationContext". And this means that you need a message for remove an observation I don't get the point.

It seems your main concern about  endpoind-id/factory #322 is about it changes the code a lot.
But it seems this will not change much more code than other PR before. And even less code than your PR #350 (even it's no truly comparable as mine is not finished)

**Comment by boaks** [comment]
I'm also not sure, if comparing the number of changed lines in incomplete PRs makes sense.
Especially, if the basic idea is already amended.

Moving the "identifier endpoint concept" into the `ObservationStore` implementation seems for me to be a perfect compromise; it enables leshan to do the explicit "endpoind-id/factory" approach while leaving californium to the "lowest change approach". It also perfectly limits the scope of the "long term endpoint identity" to the observe/notify.

**Comment by sbernard31** [comment]
So the Californium API will be : 
```java
endpoint.cancelObservation(message, token);
// or 
endpoint.cancelObservation(message);
```
This makes not so much sense to me ...
```java
endpoint.cancelObservation(endpointIdentifier, token);
```
This seems clearer.

The endpoint identifier could also help to propose an understandable API about ensuring that a request is well sent to a given peer as explained [here](https://github.com/eclipse/californium/issues/173#issuecomment-289025960)

**Comment by boaks** [comment]
The "endpoint identifier" is very abstract (not to say undefined, except, if it's based on the principal), while the `Message` rely on protocol artifacts).

I don't see the benefit of introducing such a "ep-identity/factory" approach, because I don't see the usage of the "long term endpoint identity" except for the notifies (may be until a RFC specifies something).

About issue #173:
I think, when this discussion was started, the main issue was the use of the "system unique token only". Currently my understanding is, that, if address changes are assumed, the tokens must be unique in the system anyway. So the only point left in my opinion is, that incoming messages should not only rely on that unique token and therefore should use the principal, if available, to ensure, that the message is proper related to the token (and not introduced in the system by an attacker).

**Comment by sbernard31** [comment]
> The "endpoint identifier" is very abstract (not to say undefined, except, if it's based on the principal), while the Message rely on protocol artifacts).

I disagree, you can use principal or sessionID or Session ID + CIPHER or peerAddress or DTLS connection + EPOCH. We discuss about that since 1 years and nobody seems to agree. FMPOV, most of the time principal is the best choice (if you don't allow DTLS renegociation), but this does not cover all the use cases.

> I don't see the benefit of introducing such a "ep-identity/factory" approach, because I don't see the usage of the "long term endpoint identity" except for the notifies (may be until a RFC specifies something).

Using an abstraction will make the code and the API clearer. And the factory make the code more flexible.

Again you can do without that, but your proposition with message as identifier make the API incomprehensible. Why should I pass a message to cancel an observation ? Does it mean I need to serialize the message to be able to cancel the observation later or I should create a "fake" message to cancel an observation oO ?
This seems clearer to say "I want to cancel the observation identify by the given token for the peer identified by this identifier".

**Comment by boaks** [comment]
> I disagree, you can use principal or sessionID or Session ID + CIPHER or peerAddress or DTLS connection + EPOCH. We discuss about that since 1 years and nobody seems to agree. FMPOV, most of the time principal is the best choice (if you don't allow DTLS renegociation), but this does not cover all the use cases.

I thought, we have agreed, to priorize to solve the "long term identities" used by observe/notify. So I'm not sure, how the other combinations are too useful as "long term identity".

> Using an abstraction will make the code and the API clearer. And the factory make the code more flexible. 

Only if all use case are proper defined and documented. Including the rules for extensions and/or alternative implementations.

> Why should I pass a message to cancel an observation ? 

Why should I pass  a CorrelationContext (which fields?) and a InetSocketAddress to 

`EndpointIdentifier extractIdentifier(CorrelationContext context, InetSocketAddress address);`

What must the CorrelationContext contain? What happens, if the address has changed?
Or do I create `EndpointIdentifier` depending on the used factory also on my own?

> Does it mean I need to serialize the message to be able to cancel the observation later or I should create a "fake" message to cancel an observation oO ?

Does it mean I need to serialize the correlation context to be able to cancel the observation later?
Or the "self created" `EndpointIdentifier`?

The answers to all those questions are the same for both solutions:
Just initialize the fields your `ObservationStore` requires to identify the "long term endpoint", if you just want to cancel a observe. If you want to establish an observation, you require a `Message`, if you want to process a notify, you have a `Message`.

So just see the elegance of my idea:
You can show in leshan's `ObservationStore` implementation, that your design is superior and that the flexibility really pays-off!

**Comment by boaks** [comment]
And just to repeat my opinion about the token in
https://github.com/eclipse/californium/issues/344#issuecomment-314108824

If the tokens are unique in the system, the endpoint identity may only be used for incoming Messages. If the "host/peer" want's to operate on it's own observes, it should be possible with the "system unique token" only.

**Comment by sbernard31** [comment]
> I thought, we have agreed, to priorize to solve the "long term identities" used by observe/notify. So I'm not sure, how the other combinations are too useful as "long term identity".

If you don't have authentication and changing IP environment the longer identifier you can use is SESSION ID.
If you don't have authentication and use CID (from fossati) you can use DTLS connection as identifier.
There also case where you have fixed IP.

> Only if all use case are proper defined and documented. Including the rules for extensions and/or alternative implementations.

One purpose to be flexible is to handle use case you didn't anticipate. So defined all use cases.

>Why should I pass a CorrelationContext (which fields?) and a InetSocketAddress to

    EndpointIdentifier extractIdentifier(CorrelationContext context, InetSocketAddress address);

You seems to say that this API is not so easy to understand.
Firstable, this is not a first level API, I mean this is for advanced user. So It's normal to me it's a bit harder than canceling an observation.
Secondly, it's not so easy but not so hard either. I mean if you want to customize that, this means that you understand that the default identifier is a problem for you.You see what is available in the Connection Context + InetSocketAddress and choose what match better your use case . Not so eccentric.

>The answers to all those questions are the same for both solutions:
Just initialize the fields your ObservationStore requires to identify the "long term endpoint", if you just want to cancel a observe. 

So you look at  the observationStore code, you search which fields is used in `Message`, then you create a fake `Message` with only the necessary fields. And you find the API above too complex ?

> So just see the elegance of my idea:
You can show in leshan's ObservationStore implementation, that your design is superior and that the flexibility really pays-off!

It's a question or an advice ? I didn't understand at all.
(by the way it's just not about flexibility, it's also about exposed API)

> If the tokens are unique in the system, the endpoint identity may only be used for incoming Messages. If the "host/peer" wants to operate on it's own observes, it should be possible with the "system unique token" only.

I feel that we restart the #173 discussion ...
If we chose to have unique token though all the system for out-coming messages, that means that's we should use a reserve/release tokenProvider which can be problematic in cluster environment.
And this solution does not really help as you need anyway a kind of security identifier to be sure that response is well emit by the peer you expect.
More, without endpoint identifier we need another way to express that you want to send a Request to specific peer :
```java
// create request
Request request = new Request(Code.GET);
request.setURI("my/path");
request.setDestinationIdentifier(new PSKIdentifier(pskIdentity));
request.setDestination(destinationAddress)

// send request
endpoint.sendRequest(request); 
```

**Comment by boaks** [comment]
>     So just see the elegance of my idea:
>     You can show in leshan's ObservationStore implementation, that your design is superior and that the flexibility really pays-off!
> 

> It's a question or an advice ? I didn't understand at all.

It should be a statement. I think, there is mainly a `ObservationStoreAdapter` required, which implements the "Message" base `ObservationStore` API and adapts this to the `EndpointIdentifier` / factory API approach. So it is possible for you to evolve your API design within leshan's `ObservationStore` implementation. And I feel, that this is an elegant idea, that you can implement your design without having to convince me to include a API in californium, I (currently) don't understand. If there is any information missing in the `Messages`, e.g. the `Principal` for an initial outgoing `Message`, we can add that to a (available) callback.

**Comment by boaks** [comment]
> More, without endpoint identifier we need another way to express that you want to send a Request to specific peer :

Sure, therefore

https://github.com/eclipse/californium/issues/344#issuecomment-314075391

> move principal from Request to Message, so it could also passed in.

**Comment by sbernard31** [comment]
> And I feel, that this is an elegant idea, that you can implement your design without having to convince me to include a API in californium

This design makes sense mainly at Californium level. I care about the Californium API too... 
Create a Fake message to be able to cancel an observation, sry it's too ugly for me.

> And I feel, that this is an elegant idea, that you can implement your design without having to convince me to include a API in californium.

Does it mean you decide alone what should or should not be integrated in Californium ? 

> move principal from Request to Message, so it could also passed in.

So how do you do for unauthenticated users ? ...

**Comment by boaks** [comment]
> Does it mean you decide alone what should or should not be integrated in Californium ?

I'm sure **not** "decide alone". That's an over-interpretation. 
Discussion is in my opinion somehow to convince someone else to agree. But if the other simple doesn't understand the consequences of the approach, this is very hard to do so. And therefore its much easier to have a more complete working approach, where you can see how the issues are solved together. This discussion is extremely hard to follow. And so it would be much easier, if we have a more complete working solution.

**Comment by sbernard31** [comment]
> Discussion is in my opinion somehow to convince someone else to agree.

For me is much about searching an acceptable solution together. I'm transparent if there is problem in a the solution I find better,  I don't hide it. I'm not a publicist. I'm not trying to convince you to buy washing powder.

> And so it would be much easier, if we have a more complete working solution.

That's why I started to work on a solution #322. I try to explain/share my work bit by bit  and also provide samples of how this API could be used (https://github.com/eclipse/californium/issues/173#issuecomment-289025960).

But as you seem hotly against the idea, I stopped my work to start to defend that Idea which seems to me the best one we have in stock for now.

By the way this is not my Idea but initially a Kai's idea => https://github.com/eclipse/californium/issues/173#issuecomment-288708344

**Comment by boaks** [comment]
As state multiple times:
I don't understand your solution and its consequences!
Therefore I try to make it possible for you to evolve your implementation. The proposal in 

https://github.com/eclipse/californium/issues/344#issuecomment-314075391

enables you to do so for the observations. In my opinion, this would isolate the required changes.

>     move principal from Request to Message, so it could also passed in.
> 
> So how do you do for unauthenticated peers ? ...

If the `EndpointIdentifier` is not passed into the `Connector` (or do you plan, to pass it in?), 
how will it be mapped to the information passed to the Connector? If required, we can also add additional information to all outgoing messages as I said above, 

>  If there is any information missing in the Messages, e.g. the Principal for an initial outgoing Message, we can add that to a (available) callback.  

So, if something is missing, I guess it could be easily added to the API without introducing an new concept.

About the cancel:
Sorry, may be my writing is unclear:
> If the tokens are unique in the system, the endpoint identity may only be used for incoming  Messages. If the "host/peer" wants to operate on it's own observes, it should be possible with the "system unique token" only.

In my opinion, when a Message is received, a second artefact is required to ensure the trust. This artefact is based on the received Message, either directly or by extracted into an `EndpointIdentifier`. It should be easy to do that extraction in a `ObservationStoreAdapter`.
If the cancel itself is a internal operation, FMPOV the cancel could be executed using the token only. If the cancel is based on a `Message` (RST), then we fall into the above possibilities. So I don't see, that a `fake Message` is required.

**Comment by sbernard31** [comment]
> I don't understand your solution and its consequences!
> Therefore I try to make it possible for you to evolve your implementation

Evolve to solve which issue ? You say "you don't understand" but you know it's not ok and you want to evolve it. Understanding seems to me a prerequisite before proposing changes.

> this would isolate the required changes

Why this obsession for "isolating changes" suddenly ? We are in the development branch it's normal to change the code. If a new concept is needed we must not avoid its introduction because this changes some code. (In fact not so much)

> If the EndpointIdentifier is not passed into the Connector (or do you plan, to pass it in?),
how will it be mapped to the information passed to the Connector? 

Using : `CorrelationContext createContext(EndpointIdentifier identifier)` in  `IdentifierFactory`.

>If the cancel itself is a internal operation, FMPOV the cancel could be executed using the token only. 
Cancel is not an internal operation it's public API : `CoapEndpoint.cancelObservation(byte[] token);`
Endpoint.can
I already answer to this question ... 
_If we chose to have unique token though all the system for out-coming messages, that means that's we should use a reserve/release tokenProvider which can be problematic in cluster environment.
And this solution does not really help as you need anyway a kind of security identifier to be sure that response is well emit by the peer you expect.
More, without endpoint identifier we need another way to express that you want to send a Request to specific peer :_
```java
// create request
Request request = new Request(Code.GET);
request.setURI("my/path");
request.setDestinationIdentifier(new PSKIdentifier(pskIdentity));
request.setDestination(destinationAddress)

// send request
endpoint.sendRequest(request); 
```

**Comment by boaks** [comment]
With "internal operation" I intended a operation triggered internal from the host itself, 
while "external operation" I would see a operation triggered by a received message.

I'm not sure, if it's saver for a "internal operation" to require a provided endpoint together with the unique token. I would say, no.

**Comment by sbernard31** [comment]
The endpoint identifier idea seems to not convince you.
As there is many people we seems to be waiting for Californium 2.0.0 and Leshan 1.0.0. Let this "endpoint identifier" aside.

And let's go for your approach to solve  #173, #104. As soon as code will be available I will try to integrate it in Leshan.
